### PR TITLE
[JENKINS-36240] Option to look up trust status directly from GitHub API

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-36240](https://issues.jenkins-ci.org/browse/JENKINS-36240) 

Replaces #96

To Do

- [x] Fix the NPE when calling getTrustedRevision by populating the GHRepository and opening a connection to GitHub

@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait.java [hunk 1]
```diff
@@ -316,6 +318,94 @@ public boolean isApplicableToOrigin(@NonNull Class<? extends SCMHeadOrigin> orig
...
+            extends SCMHeadAuthority<GitHubSCMSourceRequest, PullRequestSCMHead, PullRequestSCMRevision> {
+        @NonNull
+        private final GHPermissionType permission;
+
+        /**
+         * Constructor.
+         */
+        @DataBoundConstructor
+        public TrustPermission(@NonNull String permission) {
+            GHPermissionType permissionType = GHPermissionType.ADMIN;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait.java]: Why is this even configurable? The only useful value is `WRITE`. If you can push to the repository, you could change the `Jenkinsfile`, thus you are trusted.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubPermissionsSource.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.io.IOException;
+import org.kohsuke.github.GHPermissionType;
+
+/**
+ * A deferred source of permission information about a repository.
+ *
+ * @since TODO
+ */
+public abstract class GitHubPermissionsSource {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubPermissionsSource.java]: Why does this need to be `public`? Seems like an implementation detail.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSourceRequest.java [hunk 3]
```diff
@@ -113,6 +116,21 @@
      */
     @CheckForNull
     private GitHub gitHub;
+    /**
+     * The repository.
+     */
+    @CheckForNull
+    private GHRepository repository;
+    /**
+     * The resolved permissions keyed by user.
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSourceRequest.java]: How long is this cached?

**Comment by jglick** [comment]
Desirable to have https://github.com/kohsuke/github-api/pull/358 but probably unnecessary.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSourceRequest.java (hunk 3 above)
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSourceRequest.java]: For the duration of the request to scm api

File: src/main/java/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait.java (hunk 1 above)
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait.java]: Admin is also useful

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubPermissionsSource.java (hunk 2 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubPermissionsSource.java]: What `SCMSource` would that be? Best to not expose a `public` API without a concrete use case.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait.java (hunk 1 above)
Comment (abayer) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait.java]: I tend to agree with @jglick. For these purposes, `WRITE` is just as dangerous as `ADMIN`, so why do we care about `ADMIN`?

**Comment by michaelneale** [comment]
there are some findbugs problems with this - but looks close. This is a desired feature, a subtle one but I have seen people rage over it when they realise what the problem is with trust.

**Pull request merged**